### PR TITLE
kernel/timeout: Cleanup/speedup parallel announce logic

### LINKED ISSUE #44758: intel_adsp: kernel.common tests are failing
**Describe the bug**
kernel.common tests are failing on intel_adsp25/18/15 .

**To Reproduce**
1. twister -W -p intel_adsp_cavs25 --device-testing --device-serial-pty=$HOME/zephyrproject/zephyr/soc/xtensa/intel_adsp/tools/cavstwist.sh --west-flash=$HOME/zephyrproject/zephyr/soc/xtensa/intel_adsp/tools/cavstwist.sh -T tests/kernel/common/ -vvv
2. See error

Tests started failing after the commit:  b1182bf8 got merged. 

**Logs and console output**
```
Running test suite common
===================================================================
START - test_bootdelay
PASS - test_bootdelay in 0.1 seconds
===================================================================
START - test_irq_offload
PASS - test_irq_offload in 0.1 seconds
===================================================================
START - test_nested_irq_offload
Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/common/src/irq_offload.c:209: nestoff_timer_fn: (nested_executed is false)
nested irq_offload did not run
```
**Environment (please complete the following information):**
 - OS: Fedora
 - Toolchain (e.g Zephyr SDK, ...): SDK 13.2
 - Commit SHA: 37e561f42e52

**Additional context**
After git bisect, tests started failing after commit:  b1182bf8.

### PR DESCRIPTION
**Pull request by andyross**
Commit b1182bf83b53 ("kernel/timeout: Serialize handler callbacks on
SMP") introduced an important fix to timeout handling on
multiprocessor systems, but it did it in a clumsy way by holding a
spinlock across the entire timeout process on all cores (everything
would have to spin until one core finished the list).  The lock also
delays any nested interrupts that might otherwise be delivered, which
breaks our nested_irq_offload case on xtensa+SMP (where contra x86,
the "synchronous" interrupt is sensitive to mask state).

Doing this right turns out not to be so hard: take the timeout lock,
check to see if someone is already iterating
(i.e. "announce_remaining" is non-zero), and if so just increment the
ticks to announce and exit.  The original cpu will then complete the
full timeout list without blocking any others longer than needed to
check the timeout state.

Fixes #44758

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

### CODE REVIEW

**Comment by andyross** [comment]
Cleanup of the "fix racing timeout announce" issue from last week that does the same thing in a better/cleaner/smaller/faster way.  It also fixes a regression that got introduced in our irq_offload_nested test case (which happens to want to trigger a nested interrupt from within a k_timer handler, which the clumsy version disallowed by holding a lock)

**Comment by andyross** [comment]
@kv2019i This showed essentially identical behavior in my rig, but it might be worth a respin through the SOF performance case on TGL that was failing earlier just to be sure this doesn't re-break something.

File: kernel/timeout.c [hunk 1]
```diff
@@ -253,12 +242,20 @@ void sys_clock_announce(int32_t ticks)
...
 	k_spinlock_key_t key = k_spin_lock(&timeout_lock);
 
+	/* We release the lock around the callbacks below, so on SMP
+	 * systems someone might be already running the loop.  Don't
+	 * race (which will cause paralllel execution of "sequential"
+	 * timeouts and confuse apps), just increment the tick count
+	 * and return.
+	 */
+	if (IS_ENABLED(CONFIG_SMP) && announce_remaining != 0) {
+		announce_remaining += ticks;
```
Comment (dcpleung) [inline kernel/timeout.c:254]: Wonder if we should use `atomic_add()` here.

Comment (andyross) [reply kernel/timeout.c:254]: We're inside a critical section, so no need.  The spinlock is already using the atomic API, which uses a SEQ_CST builtin that will emit needed barrier instructions on architectures that require it.  There's a PR out there (maybe already merged?) that clarifies that Zephyr atomics are guaranteed to provide memory ordering.

**Review by dcpleung — APPROVED** [review]

File: kernel/timeout.c (hunk 1 above)
Comment (teburd) [reply kernel/timeout.c:254]: I considered the same thing myself so perhaps a comment here is worth adding noting atomics aren't needed, that this is already done in a critical section with what amounts to a memory barrier.

**Review by teburd — APPROVED** [review]

**Comment by teburd** [comment]
I tested it out with an independent little test that I initially stumbled upon a work around for the initial fix that caused this breakage, seems like this is still absolutely fine and I like this fix even more than the previous super timer announce loop lock :+1:

**Comment by kv2019i** [comment]
@andyross @teburd Sorry. I was a bit late to the review. I'm afraid I'm seeing a regression with this:

With this patch merged, timing observed at k_timer callback (expected avg 38400):
[    10294255.442708] (         811.692688) c3 zll-schedule       ......../zephyr_domain.c:124  INFO domain tick avg 38666, max 39485
[    11325542.057292] (         873.098938) c2 zll-schedule       ......../zephyr_domain.c:124  INFO domain tick avg 38673, max 39453
[    12357414.635417] (         832.187500) c2 zll-schedule       ......../zephyr_domain.c:124  INFO domain tick avg 38695, max 39475
[    13389260.286458] (         841.614563) c3 zll-schedule       ......../zephyr_domain.c:124  INFO domain tick avg 38694, max 39492

Before this patch I get:
[    71680153.645833] (         849.895813) c3 zll-schedule       ......../zephyr_domain.c:124  INFO domain tick avg 38400, max 38448
[    72704153.619792] (         849.401062) c3 zll-schedule       ......../zephyr_domain.c:124  INFO domain tick avg 38400, max 38445
[    73728153.489583] (         851.562500) c3 zll-schedule       ......../zephyr_domain.c:124  INFO domain tick avg 38399, max 38451
[    74752153.645833] (         846.484375) c3 zll-schedule       ......../zephyr_domain.c:124  INFO domain tick avg 38400, max 38445
 
So both max and avg are worse with this patch.

**Pull request merged**